src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+        if let Some((usage_page, usage)) = usages.next() {
+            infos.push(DeviceInfo {
+                usage_page,
+                usage,
+                ..info
+            });
+
+            // Now we can create DeviceInfo for all the other usages
+            for (usage_page, usage) in usages {
+                let prev = infos.last().unwrap();
+
+                infos.push(DeviceInfo {
+                    usage_page,
+                    usage,
+                    ..prev.clone()
+                })
+            }
+        }
+    } else {
+        infos.push(info);
+    }
+
+    Some(infos)
+}
+
+/// Fill in the extra information that's available for a USB device.
+fn fill_in_usb(device: &udev::Device, info: DeviceInfo, name: &OsStr) -> DeviceInfo {
+    let usb_dev = match device.parent_with_subsystem_devtype("usb", "usb_device") {
+        Ok(Some(dev)) => dev,
+        Ok(None) | Err(_) => {
+            return DeviceInfo {
+                manufacturer_string: WcharString::String("".into()),
+                product_string: osstring_to_string(name.into()),
+                ..info
+            }
+        }
+    };
+    let manufacturer_string = attribute_as_wchar(&usb_dev, "manufacturer");
+    let product_string = attribute_as_wchar(&usb_dev, "product");
+    let release_number = attribute_as_u16(&usb_dev, "bcdDevice");
+    let interface_number = device
+        .parent_with_subsystem_devtype("usb", "usb_interface")
+        .ok()
+        .flatten()
+        .map(|ref dev| attribute_as_i32(dev, "bInterfaceNumber"))
+        .unwrap_or(-1);
+
+    DeviceInfo {
+        release_number,
+        manufacturer_string,
+        product_string,
+        interface_number,
+        ..info
+    }
+}
+
+// From linux/hidraw.h
+struct HidrawReportDescriptor {
+    size: u32,
+    value: [u8; HID_MAX_DESCRIPTOR_SIZE],
+}
+
+impl Default for HidrawReportDescriptor {
+    fn default() -> Self {
+        Self {
+            size: 0,
+            value: [0u8; HID_MAX_DESCRIPTOR_SIZE],
+        }
+    }
+}
+
+impl HidrawReportDescriptor {
+    /// Open and parse given the "base" sysfs of the device
+    pub fn from_syspath(syspath: &Path) -> HidResult<Self> {
+        let mut descriptor = HidrawReportDescriptor::default();
+
+        let path = syspath.join("device/report_descriptor");
+        let mut f = File::open(path)?;
+        let len = f.read(&mut descriptor.value)?;
+        descriptor.size = len as u32;
+
+        Ok(descriptor)
+    }
+
+    /// Create a descriptor from a slice
+    ///
+    /// It returns an error if the value slice is too large for it to be a HID
+    /// descriptor
+    #[cfg_attr(not(test), allow(dead_code))]
+    pub fn from_slice(value: &[u8]) -> HidResult<Self> {
+        let size: u32 = match value.len().try_into() {
+            Ok(v) => v,
+            Err(_) => {
+                return Err(HidError::HidApiError {
+                    message: "HID report descriptor over 4kB".into(),
```

Comment:
Not sure why there is a 4kb error message here. The try_into::<u32>() will be able  to convert much larger values.

Reply:
yep this panics: 

```rust
    #[test]
    fn overflow() {
        let data = vec![0u8; 10 * 4096];
        let desc = HidrawReportDescriptor::from_slice(&data).expect("bla");
    }
}
```

Reply:
4kB is the max size of a HID report descriptor from sysfs, at last according to C hidapi. If something makes us read a file that's bigger, that's  probably a bug somewhere. That said, I think I missed out on an `if ` in the match or something because that error message definitely doesn't match the actual error.

Reply:
This went away with the other changes around the cursor.